src/jrd/tra.cpp

```diff
@@ -3560,8 +3560,8 @@ void jrd_tra::rollbackToSavepoint(thread_db* tdbb, SLONG number)
  **************************************/
 {
 	// merge all but one folowing savepoints into one
-	while(tra_save_point && tra_save_point->sav_next &&
-		  tra_save_point->sav_next->sav_number >= number)
+	while(tra_save_point && tra_save_point->sav_number > number &&
+		  tra_save_point->sav_next && tra_save_point->sav_next->sav_number >= number)
 	{
```

Comment:
1) AFAIU, you assume here that only one savepoint in the stack may have its number out of order, supposedly this is the looper-generated one. Couldn't there be more? E.g. started by BlockNode or SavePointNode?
2) Shouldn't the same check be applied to other places like BlockNode::execute()? I see there the kissing cousin of the loop you've fixed in tra.cpp.